src/main/java/io/appium/mitmproxy/MitmproxyJava.java

```diff
@@ -10,99 +12,120 @@
+        server.start();
 
         // python script file is zipped inside our jar. extract it into a temporary file.
         String pythonScriptPath = extractPythonScriptToFile();
 
+        final List<String> mitmproxyStartParams = new ArrayList<>();
+        mitmproxyStartParams.add(mitmproxyPath);
+        mitmproxyStartParams.add("--anticache");
+        mitmproxyStartParams.add("-p");
+        mitmproxyStartParams.add(String.valueOf(proxyPort));
+        mitmproxyStartParams.add("-s");
+        mitmproxyStartParams.add(pythonScriptPath);
+
+        // adding params if needed for mitmproxy
+        if (isNotEmpty(this.extraMitmdumpParams)) {
+            mitmproxyStartParams.addAll(this.extraMitmdumpParams);
+        }
+
         mitmproxyProcess = new ProcessExecutor()
-                .command(mitmproxyPath, "--anticache", "-s", pythonScriptPath)
+                .command(mitmproxyStartParams)
                 .redirectOutput(Slf4jStream.ofCaller().asInfo())
                 .destroyOnExit()
                 .start()
                 .getFuture();
 
-        waitForPortToBeInUse(8080);
-        System.out.println("mitmproxy started on port 8080");
-
+        waitForPortToBeInUse(proxyPort);
+        log.info("Mitmproxy started on port {}", proxyPort);
     }
 
-    private String extractPythonScriptToFile() throws URISyntaxException, IOException {
+    private String extractPythonScriptToFile() throws IOException {
         File outfile = File.createTempFile("mitmproxy-python-plugin", ".py");
 
-        InputStream instream = getClass().getClassLoader().getResourceAsStream("scripts/proxy.py");
-        FileOutputStream outstream = new FileOutputStream(outfile);
-
-        byte[] buffer = new byte[1024];
+        try (
+                InputStream inputStream = getClass().getClassLoader().getResourceAsStream("scripts/proxy.py");
+                FileOutputStream outputStream = new FileOutputStream(outfile)) {
 
-        int length;
-        /*copying the contents from input stream to
-         * output stream using read and write methods
-         */
-        while ((length = instream.read(buffer)) > 0){
-            outstream.write(buffer, 0, length);
+            IOUtils.copy(inputStream, outputStream);
         }
 
-        //Closing the input/output file streams
-        instream.close();
-        outstream.close();
-
         return outfile.getCanonicalPath();
     }
 
-    public void stop() throws IOException, InterruptedException {
+    public void stop() throws InterruptedException {
         if (mitmproxyProcess != null) {
             mitmproxyProcess.cancel(true);
         }
         server.stop(1000);
-        Thread.sleep(200); // this pains me. but it seems that it takes a moment for the server to actually relinquish the port it uses.
+        waitForPortToBeFree(proxyPort);
+    }
+
+    private void waitForPortToBeFree(int port) {
+
+        while (true) {
```

Comment:
can we add a timeout here to prevent endless loop? even 5 minutes would be enough

Reply:
yes